ksqldb-rest-app/src/test/java/io/confluent/ksql/api/TlsTest.java

```diff
@@ -49,18 +61,82 @@ protected KsqlRestConfig createServerConfig() {
     config.put(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG, trustStorePassword);
     config.put(KsqlRestConfig.VERTICLE_INSTANCES, 4);
 
+    config.put(KsqlRestConfig.SSL_KEYSTORE_RELOAD_CONFIG, true);
+
     return new KsqlRestConfig(config);
   }
 
   @Override
   protected WebClientOptions createClientOptions() {
+    // for this test file, the client must use a different trust store location than the server
+    // since the client store should always be valid even when the server store is loaded with an
+    // invalid cert
+    String clientTrustStorePath = ServerKeyStore.clientKeyStoreProps()
+        .get(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG);
+    String clientTrustStorePassword = ServerKeyStore.clientKeyStoreProps()
+        .get(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG);
+
     return new WebClientOptions().setSsl(true).
         setUseAlpn(true).
         setProtocolVersion(HttpVersion.HTTP_2).
-        setTrustAll(true).
+        setTrustStoreOptions(
+            new JksOptions().setPath(clientTrustStorePath).setPassword(clientTrustStorePassword)).
         setVerifyHost(false).
         setDefaultHost("localhost").
         setDefaultPort(server.getListeners().get(0).getPort());
   }
 
+  @Test
+  public void shouldReloadCert() throws Exception {
+    JsonObject requestBody = new JsonObject().put("sql", DEFAULT_PULL_QUERY);
+
+    // Given: sanity check that a query succeeds
+    HttpResponse<Buffer> response = sendRequest("/query-stream", requestBody.toBuffer());
+    assertThat(response.statusCode(), is(200));
+    assertThat(response.statusMessage(), is("OK"));
+
+    try {
+      // When: load expired key store
+      ServerKeyStore.loadExpiredServerKeyStore();
+      assertThatEventually(
+          "Should fail to execute query with expired key store",
+          () -> {
+            // re-create client since server port changes on restart
+            this.client = createClient();
+
+            try {
+              // this should fail
+              sendRequest("/query-stream", requestBody.toBuffer());
+              return "error: request should have failed but did not";
+            } catch (Exception e) {
+              assertThat(e,
+                  instanceOf(ExecutionException.class)); // thrown from CompletableFuture.get()
+              return e.getMessage();
+            }
+          },
+          containsString("javax.net.ssl.SSLHandshakeException: Failed to create SSL connection"),
+          TimeUnit.SECONDS.toMillis(1),
+          TimeUnit.SECONDS.toMillis(1)
+      );
+    } finally { // restore cert regardless of failure above so as to not affect other tests
+      // When: load valid store
+      ServerKeyStore.loadValidServerKeyStore();
```

Comment:
I think it would be good to test reloading the server more than once - i.e. changing the cert more than once, as there might be a bug in the filewatcher/reloading that only manifests after the reload has been done once.

Reply:
The test already reloads the server twice: once with a bad cert, and then again with a good cert. Are you suggesting we should repeat the cycle twice, for a total of four reloads?

Reply:
Ah, if you're already reloading more than once that seems fine :)